sycl_runtime/lib/sycl_runtime.cpp

```diff
@@ -0,0 +1,260 @@
+                           nullptr};
+  auto ze_device = sycl::get_native<sycl::backend::ext_oneapi_level_zero>(
+      sycl_queue.get_device());
+  auto ze_ctx = sycl::get_native<sycl::backend::ext_oneapi_level_zero>(
+      sycl_queue.get_context());
+  L0_SAFE_CALL(zeModuleCreate(ze_ctx, ze_device, &desc, &ze_module, nullptr));
+  return ze_module;
+}
+
+void *getKernel(void *queue, void *module, const char *name) {
+  assert(module);
+  assert(name);
+  auto sycl_queue = static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_;
+  ze_module_handle_t ze_module = static_cast<ze_module_handle_t>(module);
+  ze_kernel_handle_t ze_kernel;
+  sycl::kernel *sycl_kernel;
+  ze_kernel_desc_t desc = {};
+  desc.pKernelName = name;
+  L0_SAFE_CALL(zeKernelCreate(ze_module, &desc, &ze_kernel));
+
+  sycl::kernel_bundle<sycl::bundle_state::executable> kernel_bundle =
+      sycl::make_kernel_bundle<sycl::backend::ext_oneapi_level_zero,
+                               sycl::bundle_state::executable>(
+          {ze_module}, sycl_queue.get_context());
+
+  auto kernel = sycl::make_kernel<sycl::backend::ext_oneapi_level_zero>(
+      {kernel_bundle, ze_kernel}, sycl_queue.get_context());
+  sycl_kernel = new sycl::kernel(kernel);
+  return sycl_kernel;
+}
+
+void launchKernel(void *queue, sycl::kernel *kernel, size_t gridX, size_t gridY,
+                  size_t gridZ, size_t blockX, size_t blockY, size_t blockZ,
+                  size_t sharedMemBytes, ParamDesc *params) {
+  auto sycl_queue = static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_;
+  auto sycl_global_range =
+      ::sycl::range<3>(blockZ * gridZ, blockY * gridY, blockX * gridX);
+  auto sycl_local_range = ::sycl::range<3>(blockZ, blockY, blockX);
+  sycl::nd_range<3> sycl_nd_range(
+      sycl::nd_range<3>(sycl_global_range, sycl_local_range));
+
+  auto paramsCount = countUntil(params, ParamDesc{nullptr, 0});
+
+  sycl_queue.submit([&](sycl::handler &cgh) {
+    for (size_t i = 0; i < paramsCount; i++) {
+      auto param = params[i];
+      cgh.set_arg(static_cast<uint32_t>(i),
+                  *(static_cast<void **>(param.data)));
+    }
+    cgh.parallel_for(sycl_nd_range, *kernel);
+  });
+}
+
+// Wrappers
+
+extern "C" SYCL_RUNTIME_EXPORT void *gpuCreateStream(void *device,
+                                                     void *context) {
+  return catchAll([&]() {
+    if (!device && !context) {
+      return new GPU_SYCL_QUEUE();
+    } else if (device && context) {
+      return new GPU_SYCL_QUEUE(static_cast<sycl::device *>(device),
+                                static_cast<sycl::context *>(context));
+    } else if (device && !context) {
+      return new GPU_SYCL_QUEUE(static_cast<sycl::device *>(device));
+    } else {
+      return new GPU_SYCL_QUEUE(static_cast<sycl::context *>(context));
+    }
+  });
+}
+
+extern "C" SYCL_RUNTIME_EXPORT void gpuStreamDestroy(void *queue) {
+  catchAll([&]() { delete static_cast<GPU_SYCL_QUEUE *>(queue); });
```

Comment:
When this object is being deleted., do you have to notify SYCL runtime you are not using these handles hold by this object?  like those   sycl::device sycl_device_;
  sycl::context sycl_context_;
  sycl::queue sycl_queue_;

Reply:
no, I don't need to free it, when I delete the struct, all its members are deleted. Their scope is as much as the scope of the class or struct they belong to.

Reply:
Sounds good. For runtime module, ususally people create focus test to see whether it is leaking memory. Please considering add one to capture potential memory leak (sometime leaking may happen in the user level driver, so monitoring the virtual memory used by process could capture this kind of bug.